sksurv/tree/_criterion.pyx

```diff
@@ -328,12 +328,29 @@ cdef class LogrankCriterion(Criterion):
             DOUBLE_t ratio
             DOUBLE_t n_events
             DOUBLE_t n_at_risk
+            DOUBLE_t dest_j0
+            DOUBLE_t dest_j1
 
         self.riskset_total.at(0, &n_at_risk, &n_events)
         ratio = n_events / n_at_risk
         dest[0] = ratio  # Nelson-Aalen estimator
         dest[1] = 1.0 - ratio  # Kaplan-Meier estimator
 
+        # low memory mode
+        if  self.n_outputs == 1:
+            dest_j0 = dest[0]
+            dest_j1 = dest[1]
+            for i in range(1, self.n_unique_times):
+                self.riskset_total.at(i, &n_at_risk, &n_events)
+                if n_at_risk != 0:
+                    ratio = n_events / n_at_risk
+                    dest_j0 += ratio
+                    dest_j1 *= 1.0 - ratio
+                if True: # TODO: only sum for event times
+                    dest[0] += dest_j0
+                    dest[1] += dest_j1
```

Comment:
The sum over the survival function can be skipped too, the sum over the CHF (`dest_j0`) is sufficient.

If the `i`-th unique time does not have an event, you would not add anything to the sum. The corresponding Python code is https://github.com/sebp/scikit-survival/blob/bd2e24029632c4a0dbccff86ee905ba8d9b6f5e7/sksurv/tree/tree.py#L368
where `is_event_time` is created at https://github.com/sebp/scikit-survival/blob/bd2e24029632c4a0dbccff86ee905ba8d9b6f5e7/sksurv/tree/tree.py#L215
It is currently not accessible from Cython, so you would need to pass it to `LogrankCriterion`.